rhel/vex/parser.go

```diff
@@ -0,0 +1,471 @@
+package vex
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/klauspost/compress/snappy"
+	"github.com/package-url/packageurl-go"
+	"github.com/quay/zlog"
+
+	"github.com/quay/claircore"
+	"github.com/quay/claircore/rhel/internal/common"
+	"github.com/quay/claircore/toolkit/types/cpe"
+	"github.com/quay/claircore/toolkit/types/csaf"
+	"github.com/quay/claircore/toolkit/types/cvss"
+)
+
+// Parse implements [driver.Updater].
+func (u *Updater) Parse(ctx context.Context, contents io.ReadCloser) ([]*claircore.Vulnerability, error) {
+	// NOOP
+	return nil, errors.ErrUnsupported
+}
+
+// DeltaParse implements [driver.DeltaUpdater].
+func (u *Updater) DeltaParse(ctx context.Context, contents io.ReadCloser) ([]*claircore.Vulnerability, []string, error) {
+	ctx = zlog.ContextWithValues(ctx, "component", "rhel/vex/Updater.DeltaParse")
+	// This map is needed for deduplication purposes, the compressed CSAF data maybe include
+	// entries that have been subsequently updated in the changes.
+	out := map[string][]*claircore.Vulnerability{}
+	deleted := []string{}
+
+	pc := newProductCache()
+	rc := newRepoCache()
+
+	r := bufio.NewReader(snappy.NewReader(contents))
+	for b, err := r.ReadBytes('\n'); err == nil; b, err = r.ReadBytes('\n') {
+		c, err := csaf.Parse(bytes.NewReader(b))
+		if err != nil {
+			return nil, nil, fmt.Errorf("error parsing CSAF: %w", err)
+		}
+		name := c.Document.Tracking.ID
+		if c.Document.Tracking.Status == "deleted" {
+			deleted = append(deleted, name)
+			continue
+		}
+
+		var selfLink string
+		for _, r := range c.Document.References {
+			if r.Category == "self" {
+				selfLink = r.URL
+			}
+		}
+		ctx = zlog.ContextWithValues(ctx, "link", selfLink)
+		creator := newCreator(name, selfLink, c, pc, rc)
+		for _, v := range c.Vulnerabilities {
+			// Create vuln here, there should always be one vulnerability
+			// here in the case of RH VEX but the spec allows multiple.
+			links := []string{}
+			for _, r := range v.References {
+				links = append(links, r.URL)
+			}
```

Comment:
It'd be nice to ensure "self" is first (this one. Not the one from `c.Document.References`) so implementations who just look at the first link will always get this one. I know StackRox does this, but I'm not sure if Quay does, too

Reply:
You think a user would prefer to parse through a big blob of JSON rather than the RH CVE page?

Reply:
I'm referring to this "self" under `vulnerabilities.references`:

```
"references": [
        {
          "category": "self",
          "summary": "Canonical URL",
          "url": "https://access.redhat.com/security/cve/CVE-2024-3727"
        },
        {
          "category": "external",
          "summary": "RHBZ#2274767",
          "url": "https://bugzilla.redhat.com/show_bug.cgi?id=2274767"
        },
```

not sure if it's guaranteed to be first, but it's definitely the one users will care about most. Though, that can just be the client's problem to fetch it, if that's the one they want (for example StackRox can just search for the link prefixed with `https://access.redhat.com/security/cve/CVE-`